### PR TITLE
Clarify/Improve docs on family-names vs generic-families

### PR DESCRIPTION
**Pull request by aitikgupta**
## PR Summary
Coming from the answer at https://discourse.matplotlib.org/t/fonts-and-font-families/22132 and a brief discussion with @anntzer, this PR modifies the content of font docs (and one example, for which I have a slight niggle against - because it's not really the most obvious and best way to guide a user into setting font family).

Hopefully, we might be moving font stuff out from this [very long page](https://matplotlib.org/stable/tutorials/text/text_props.html), and probably a better example in future.

This is possibly the beginning of handling https://github.com/matplotlib/matplotlib/issues/5941.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: examples/text_labels_and_annotations/font_family_rc_sgskip.py
Comment (anntzer) [inline examples/text_labels_and_annotations/font_family_rc_sgskip.py:7]: there are 5 generic families only, you may as well list them all...

File: tutorials/text/text_props.py
Comment (anntzer) [inline tutorials/text/text_props.py:216]: I would leave out the "not recommended" unless you write another paragraph to explain the motivation of having generic families at all.

File: examples/text_labels_and_annotations/font_family_rc_sgskip.py
Comment (aitikgupta) [reply examples/text_labels_and_annotations/font_family_rc_sgskip.py:7]: 👍🏼

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py]: Added another paragraph 👍🏼

**Review by anntzer — APPROVED** [review]
Looks good to me, but perhaps someone less familiar with fonts should check whether this looks like an improvement to them :)

File: tutorials/text/text_props.py
Comment (QuLogic) [inline tutorials/text/text_props.py:204]:
```suggestion
#    # This is effectively translated to:
```

File: tutorials/text/text_props.py
Comment (QuLogic) [inline tutorials/text/text_props.py:220]: This needs some re-wording. The first paragraph and the second are not really linked together, and then the code after seems disjoint from what is being explained.

File: tutorials/text/text_props.py
Comment (QuLogic) [inline tutorials/text/text_props.py:222]: appended, or prepended?

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:198]: DId it to relate to the previous table's heading, let me remove..

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:204]: 👍🏼

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:222]: Appended, we're talking about generic family here (sans-serif in this case).

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:220]: I see, let me try rewording..

File: tutorials/text/text_props.py
Comment (tacaswell) [inline tutorials/text/text_props.py:231]: But we do not currently do per-glyph fallback, only "can we find this font?" fallback.

File: tutorials/text/text_props.py
Comment (tacaswell) [inline tutorials/text/text_props.py:221]: This is a bit confusing, because this pattern is exactly what our default values are!

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:231]: That's true, but even in-general usecases, this holds true.
(would be 100% true after we've accomplished true fallback)

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:221]: And those default values are exactly what we don't want users to touch, if they want custom font families, go ahead and modify `font.family`, what we shouldn't recommend is to modify `font.serif`, `font.fantasy`, etc.

File: tutorials/text/text_props.py
Comment (QuLogic) [reply tutorials/text/text_props.py]: Ah, I misread that.

File: tutorials/text/text_props.py
Comment (tacaswell) [reply tutorials/text/text_props.py:231]: We should hold off on including "...contain the required glyph." until it is all the way true ;)

It is my understanding that these values eventually make it to https://github.com/matplotlib/matplotlib/blob/a3fbf1876830a89087842ea779515e0d099c885d/lib/matplotlib/font_manager.py#L1130-L1158 which only takes into account the name (not the pages supported).

Comment (aitikgupta) [reply tutorials/text/text_props.py:231]: That also makes sense, let me remove that line 👍🏼

File: tutorials/text/text_props.py
Comment (tacaswell) [reply tutorials/text/text_props.py:221]: I guess there is a hypothetical workflow where people are setting `matplotlib.rcParams['font.family'] = ['serif']` or `matplotlib.rcParams['font.family'] = ['cursive']`.  People could do

```
matplotlib.rcParams['font.sans-serif'] = ['Source Han Sans TW', ...]
matplotlib.rcParams['fonts.serif'] = ['Noto Serif', ...]
```

and then switch between serif/sans serif in a straight forward way.  You could then imagine filling these lists with a number of fonts which _may_ be installed and using the first one you find.

The current default values include a bunch of fonts that we do not ship (but expect to be on users systems with high probability).  The only fonts we ship are in https://github.com/matplotlib/matplotlib/tree/master/lib/matplotlib/mpl-data/fonts/ttf (there are a bunch of afm files as well, but those are only font metrics (glyph sizes) that we need for layout, not the full fonts).

Comment (aitikgupta) [reply tutorials/text/text_props.py:221]: But this exactly creates confusions for people who're coming to our docs for setting custom fonts, like https://discourse.matplotlib.org/t/fonts-and-font-families/22132.

I'm sure an experienced developer could come up with this quick way of switching regardless of the docs. We _could_ however, add a small paragraph about where you _can_ use this type of method (giving this switching example) maybe?

It's more like saying, "You can use the non-recommended way if you know what you're doing. But if you're new to this, or if you're coming from a web-background, you should probably stick to the recommended way."

Comment (aitikgupta) [reply tutorials/text/text_props.py:221]: The default values not being shipped is a bit of concern, because the statement becomes invalid (maybe I could rephrase?)

File: tutorials/text/text_props.py
Comment (jklymak) [inline tutorials/text/text_props.py:201]: What is `MyFamily1` and `MyFamily2`?  Are these fonts or font families?  How do I get these names?

File: tutorials/text/text_props.py
Comment (jklymak) [inline tutorials/text/text_props.py:202]: ditto, what is `MySerif1/2`?    Are these font names?  Where do I get these?  Where are the fonts installed?

**Review by jklymak — DISMISSED** [review]
I think I'm ignorant enough about what is going on to say this raises more questions than it answers ;-).  Apologies if this is explained elsewhere and I just could not find it.

I also am not at all sure this belongs in a tutorial called `text_props`.  But maybe re-org can happen later?  I think the tutorial should have "font" in the title somewhere....

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:152]: All font families which not denoted by this set: `{'cursive', 'serif', 'sans-serif', 'fantasy', 'monospace'} are user-defined font families.

We basically draw a line between "generic-families" vs "family-names", see [this link](https://drafts.csswg.org/css-fonts-3/#font-family-prop):

> There are two types of font family names:
    family-name
    The name of a font family of choice such as Helvetica or Verdana in the previous example.
    generic-family
    The following generic family keywords are defined: ‘serif’, ‘sans-serif’, ‘cursive’, ‘fantasy’, and ‘monospace’.

Comment (aitikgupta) [reply tutorials/text/text_props.py:152]: Though I agree, we could use a better name than 'user-defined' font families, because it might give the impression that a user himself has to define that font.

Any suggestions?

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:201]: These are font families, like I said in the previous comment, it can be anything other than those 5 special keywords.

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:202]: Same as above^^

(we could also say that all fonts installed in user's machine is a 'user-defined' fonts)

(also note that sometimes the word 'fonts' and 'font-families' are used interchangeably)

File: tutorials/text/text_props.py
Comment (jklymak) [reply tutorials/text/text_props.py:152]:
```suggestion
# | ``'font.family'``   | List of font families (installed on the users machine) and/or |
```

File: tutorials/text/text_props.py
Comment (jklymak) [inline tutorials/text/text_props.py:178]:
```suggestion
# 'monospace', 'sans', 'sans serif', 'sans-serif', 'serif'}``) and actual font families
```

File: tutorials/text/text_props.py
Comment (jklymak) [reply tutorials/text/text_props.py:152]: Suggest before this table something like:

```
Matplotlib can use font families installed on the user's computer, i.e. Helvetica, Times, etc, as well as other user-installed fonts.  Font families can also be specified with generic names like 'serif', 'sans-serif', in which case a fallback list is used:
```

File: tutorials/text/text_props.py
Comment (jklymak) [reply tutorials/text/text_props.py:201]: So I'd change these to `Family1` or `FontFamily1`, or even just be specific with generic fonts...

Also, what do these default to?  What are the default 'serif' fonts?

Comment (aitikgupta) [reply tutorials/text/text_props.py:201]: The default "generic-family" fonts default to the ones in rcParams, mentioned in the [docs](https://matplotlib.org/stable/tutorials/introductory/customizing.html):
image

File: tutorials/text/text_props.py
Comment (jklymak) [reply tutorials/text/text_props.py:152]: So maybe here, link the rcParam so folks can see the default list?

**Comment by aitikgupta** [comment]
^reworded sentences as per @jklymak's suggestions.

> I also am not at all sure this belongs in a tutorial called text_props. But maybe re-org can happen later? I think the tutorial should have "font" in the title somewhere....

Yes, I think that's definitely something which we should consider, though we should discuss where and how exactly fonts could be abstracted/put.

File: tutorials/text/text_props.py
Comment (jklymak) [inline tutorials/text/text_props.py:152]:
```suggestion
# | ``'font.family'``   | List of font families (installed on       |
```

**Review by jklymak — COMMENTED** [review]
This seems better to me, thanks!

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:152]: 👍🏼

File: tutorials/text/text_props.py
Comment (jklymak) [inline tutorials/text/text_props.py:178]:
```suggestion
# Helvetica, Times, etc.
```

File: tutorials/text/text_props.py
Comment (tacaswell) [inline tutorials/text/text_props.py:189]:
```suggestion
# | css family alias                        | rcParam with mappings          |
```

Maybe name this as the CSS names (to head of my question of "did Matplotlib make up these names!?").

**Comment by tacaswell** [comment]
```

/home/circleci/project/doc/tutorials/text/text_props.rst:219: WARNING: undefined label: /tutorials/introductory/customizing
generating indices... genindex py-modindex done
```

These failures look real and related.

File: tutorials/text/text_props.py
Comment (jklymak) [reply tutorials/text/text_props.py:189]: Maybe `CSS-style family alias`.  Many of our users do not know what css is really.

**Comment by aitikgupta** [comment]
> These failures look real and related

Yeah even I was confused when I first saw it, it's basically coming from https://github.com/matplotlib/matplotlib/blob/5d7979be02029469f960119a602798dc7b69b975/tutorials/text/text_props.py#L184

Is there any other way we can reference this link: https://matplotlib.org/stable/tutorials/introductory/customizing.html?

File: tutorials/text/text_props.py
Comment (aitikgupta) [reply tutorials/text/text_props.py:189]: I guess `CSS-based generic-family alias` would make more sense? (special focus on the 'generic-family')

File: examples/text_labels_and_annotations/font_family_rc_sgskip.py
Comment (aitikgupta) [reply examples/text_labels_and_annotations/font_family_rc_sgskip.py:7]: I knew I missed this somewhere, thanks!

**Comment by jklymak** [comment]
Aren't those usually `:doc:`?  though I've not seen the form where you make it a link to alternate text...

**Review by jklymak — APPROVED** [review]
Anyone can merge when CI passes, though @aitikgupta you may want to squash your commits?

**Comment by aitikgupta** [comment]
I was waiting for tests to pass first, but I can squash and push right now?

**Comment by jklymak** [comment]
I don't think there is any reason for CI to run twice, unless you want to keep the commits around for some reason.

**Comment by jklymak** [comment]
Thanks @aitikgupta  - looks helpful!

**Pull request merged**
